### ISSUE TITLE
Forces make fitting worse with normalized features

### ISSUE THREAD

**Issue by kzinovjev** [lab-cosmo/librascal#427]
Hi,

In the MLIP example I was playing with different values of λ and feature normalization and noticed the following behaviour. 
I used the entire dataset (2000 structures) without splitting it into train/test sets. 

With `n_sparse = {1:40,6:30,8:20}` and `normalize=False` in `hypers` (as in the original example) and putting a very weak regularizer on forces (0.0001) I get very large systematic error (66 eV, `np.mean(y_pred - y)`). To compare, the mean deviation of the baseline (sum of all atomic self contributions) is 53 eV, so it doesn't seem that this is due to the model just falling back to baseline. Once this shift is subtracted from the predictions, the model is reasonably good (RMSE=0.1 eV). For larger force λ the model gets worse, but the systematic error gets reduced.

This behaviour disappears when I use `normalize=False`, but also I see that the forces becoming unhelpful at all to the model - I get the best predictions (even better than with `normalize=False` and weak regularizers) with force λ=1, which, as I understand, essentially means ignoring the forces. Here are the results:

| Energy λ | Force λ | normalize=False | | normalize=True | |
| --- | --- | --- | --- | --- | --- |
|  |  | mean error | RMSE* | mean error | RMSE* |
| 0.1 | 1 | 32.103 | 0.669 | 0 | 0.095 |
| 0.1 | 0.1 | 35.009 | 0.611 | 0.001 | 0.097 |
| 0.1 | 0.01 | 35.151 | 0.155 | 0.099 | 0.097 |
| 0.1 | 0.0001 | **_66.166_** | **_0.101_** | 1.278 | 0.097 |
| 0.0001 | 1 | 0 | 0.100 | 0 | **_0.060_** |
| 0.0001 | 0.1 | 0 | 0.100 | 0 | 0.061 |
| 0.0001 | 0.01 | 0 | 0.093 | 0 | 0.065 |
| 0.0001 | 0.0001 | 0.015 | 0.105 | 0.001 | 0.097 |

Then I tried using much fewer sparse points (`n_sparse = {1:20,6:15,8:10}`). The behaviour got exaggerated for `normalize=False` and I started seeing large systematic errors with an opposite sign for `normalize=True`:

| Energy λ | Force λ | normalize=False | | normalize=True | |
| --- | --- | --- | --- | --- | --- |
|  |  | mean error | RMSE* | mean error | RMSE* |
| 0.1      | 1       | 32.104          | 0.669          | -0.002        | 0.145         |
| 0.1      | 0.1     | 35.125          | 0.611          | -0.148        | 0.163         |
| 0.1      | 0.01    | 36.912          | 0.221          | -11.818       | 0.160         |
| 0.1      | 0.0001  | **_82.321_**          | **_0.144_**          | **_-57.897_**       | 0.154         |
| 0.0001   | 1       | 0               | 0.129          | 0             | **_0.099_**         |
| 0.0001   | 0.1     | 0               | 0.129          | 0             | 0.099         |
| 0.0001   | 0.01    | 0               | 0.126          | 0             | 0.105         |
| 0.0001   | 0.0001  | 0.074           | 0.154          | -0.148        | 0.164         |

So the stuff I don't get is:
1. Why with `normalize=True` the forces do not improve the fit even in case of very few sparse points (so, no overfitting)?
2. Why I get huge systematic errors with low forces regularizer, which are negative (so the predictions get farther from the baseline than the training data) in case of `normalize=True` and few sparse points?

Attached the modified notebook. [MLIP_example_lambda.zip](https://github.com/lab-cosmo/librascal/files/11202853/MLIP_example_lambda.zip)

Thanks in advance!

**Comment by agoscinski** [comment]
Maybe first some general comments on standardization of features

We had quite a few heated debates how to standardize the features. In principle, you want to standardize your features for numerical accuracy and to consider always the same range of regularizers (in the range of [0, 1]), when training your model. The relative scaling between our features is important, since they are all based on expansion coefficients from the local atomic densities. So we want to preserve this relative scaling, which means that we scale all features by a global factor. However, when this library was created we followed the current state of the art that normalized the features row by row (sample by sample) which distorts the relative scaling between samples and can produce sometimes results that are very wrong. Most of the time for larger datasets we don't really observe much difference in performance, so some people still use it. Especially, because there is no other inbuilt standardization provided from librascal. And even though your distort your relative scaling between samples, you can use the same regularizer range. Some other people in our lab use a custom Standardizer which standardizes by a global factor (https://scikit-matter.readthedocs.io/en/latest/preprocessing.html#skmatter.preprocessing.StandardFlexibleScaler), but this does not yet work with forces (there is some work on making this possible with forces, if you are interested).

So comparing `normalize=True` and `normalize=False` for the same range of regularizers is not so meaningful, since the regularizers for `normalize=False` need to be much smaller.

1. Are you showing energy errors? If you include forces you energy error might increase while your force error is decreasing, when enough data is present. Force information improves usually energy predictions, when the dataset is small. The model is overall more robust if it is fitted also on the forces, but the prediction error for energies might be higher (a bit in the spirit of https://arxiv.org/abs/1908.02729 even though the paper is about regularization of the derivatives and not learning them). I would try `normalize=False` with in the range of [1e-25, 1e-3].

2. I would say this is more related to the large energy λ which penalizes large weights, thus you cannot learn the baseline well. I don't have a good explanation, however, why this does not happen with more sparse points when normalize=True.

**Comment by kzinovjev** [comment]
Thanks for the answer! I had wrong understanding of how the forces are used for training. I thought that the derivative kernels are explicitly used during the prediction so that the derivatives of the fitted function (given that the regularization is low) are guaranteed to follow the reference derivative values. Now I see that the forces are used merely as a "regularizer" to improve the weights of the energy kernels. So the functional capacity is the same with or without using forces for the training. Please correct me if I'm wrong or if I mess up the terminology.

Now the results make more sense. When I use a much larger forces lambda (essentially not using forces) I get RMSE of 0.107 and 0.821 for energies and forces respectively, but with larger energies lambda I get 0.145 and 0.656.

My only doubt is: is it really worth using forces if I have a sufficiently large number of sparse points? The memory overhead is huge and the impact doesn't seem to be that large. Moreover, I'm mostly interested in constructing free energy profiles, so the accuracy of forces is not as important as the relative energies of different states.
